proxy/src/control/discovery.rs

```diff
@@ -369,6 +377,82 @@ where
     }
 }
 
+// ===== impl DestinationSet =====
+
+impl <T: HttpService<ResponseBody = RecvBody>> DestinationSet<T> {
+    fn add<Addrs>(&mut self, authority_for_logging: &FullyQualifiedAuthority, addrs_to_add: Addrs)
+        where Addrs: Iterator<Item = SocketAddr>
+    {
+        let mut addrs = match self.addrs.take() {
+            Exists::Yes(addrs) => addrs,
+            Exists::Unknown | Exists::No => {
+                trace!("adding entries for {:?} that wasn't known to exist. Now assuming it does.",
+                       authority_for_logging);
+                HashSet::new()
+            },
+        };
+        for addr in addrs_to_add {
+            if addrs.insert(addr) {
+                trace!("update {:?} for {:?}", addr, authority_for_logging);
+                // retain is used to drop any senders that are dead
+                self.txs.retain(|tx| {
+                    tx.unbounded_send(Update::Insert(addr)).is_ok()
+                });
+            }
+        }
+        self.addrs = Exists::Yes(addrs);
+    }
+
+    fn remove<Addrs>(&mut self, authority_for_logging: &FullyQualifiedAuthority,
+                     addrs_to_remove: Addrs)
+        where Addrs: Iterator<Item = SocketAddr>
+    {
+        let addrs = match self.addrs.take() {
+            Exists::Yes(mut addrs) => {
+                for addr in addrs_to_remove {
+                    if addrs.remove(&addr) {
+                        self.notify_of_removal(addr, authority_for_logging)
+                    }
+                }
+                addrs
+            },
+            Exists::Unknown | Exists::No => {
+                trace!("remove addresses for {:?} that wasn't known to exist. Now assuming it does.",
+                       authority_for_logging);
+                HashSet::new()
+            },
+        };
+        self.addrs = Exists::Yes(addrs)
+    }
+
+    fn no_endpoints(&mut self, authority_for_logging: &FullyQualifiedAuthority, exists: bool) {
+        trace!("no endpoints for {:?} that is known to {}", authority_for_logging,
+               if exists { "exist"} else { "not exist"});
+        match self.addrs.take() {
+            Exists::Yes(addrs) => {
```

Comment:
could just be `if let Exists::Yes(addrs) = self.addrs.take() {` since the other cases are not meaningful.

Reply:
I avoided using `if let` because I think it is really important to consider all the cases. In general I think `if let` leads to bugs because it makes it easy to ignore cases, especially when refactoring in a way that adds new cases that need to be considered.